File: web/sign_features.mjs
// Port of assets/modules/features.py. Coordinates are from the unmirrored image.
const hp = [0,1,2,3,0,5,6,7,0,9,10,11,0,13,14,15,0,17,18,19];
const hc = [1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20];
const up = [1,1,2,3,1,5,6], uc = [0,2,3,4,5,6,7];
const zero = n => Array.from({length:n}, () => [0,0,0]);
const pixel = (p,w,h) => [Math.fround(p.x*w), Math.fround(p.y*h), Math.fround(p.z*w)];

function geometry(coords, parents, children, a, b) {
  const n = parents.length, out = new Float32Array(n*3+a.length);
  if (coords.flat().every(v => v === 0)) return out;
  const bones = parents.map((p,i) => coords[children[i]].map((v,j) => Math.fround(v-coords[p][j])));
  bones.forEach(([x,y,z],i) => {
    const norm = Math.hypot(x,y)+1e-6;
    out[i]=x/norm; out[n+i]=y/norm; out[2*n+i]=z;
  });
  a.forEach((v,i) => {
    const [x,y]=bones[v], [s,t]=bones[b[i]];
    out[3*n+i]=Math.acos(Math.max(-1,Math.min(1,(x*s+y*t)/(Math.hypot(x,y)*Math.hypot(s,t)+1e-6))));
  });
  return out;
}

export function extractFeatures(hands, handedness, poseLandmarks, w, h) {
  if (!(w>0 && h>0)) throw new Error('Invalid frame dimensions');
  let pose=zero(8), right=zero(21), left=zero(21);
  if (poseLandmarks?.length) {
    const p=poseLandmarks;
    const ls=pixel(p[11],w,h), rs=pixel(p[12],w,h);
    pose=[pixel(p[0],w,h),ls.map((v,i)=>Math.fround((v+rs[i])/2)),rs,
      pixel(p[14],w,h),pixel(p[16],w,h),ls,pixel(p[13],w,h),pixel(p[15],w,h)];
  }
  hands.forEach((hand,i)=> {
    const name=handedness[i]?.[0]?.categoryName;
    if (name==='Right') right=hand.map(p=>pixel(p,w,h));
    if (name==='Left') left=hand.map(p=>pixel(p,w,h));
  });
  const neck=pose[1].slice();
  const width=Math.hypot(...pose[5].map((v,i)=>v-pose[2][i]));
  const scale=width<1e-6 ? 1 : width;
  const normalize=coords=>coords.map(p=>p.map((v,i)=>Math.fround(Math.fround(v-neck[i])/scale)));
  pose=normalize(pose);
  if (right.some(p=>p.some(v=>v!==0))) right=normalize(right);
  if (left.some(p=>p.some(v=>v!==0))) left=normalize(left);
  const out=new Float32Array(150);
  out.set(geometry(right,hp,hc,[0,4,8],[1,5,9]),0);
  out.set(geometry(left,hp,hc,[0,4,8],[1,5,9]),63);
  out.set(geometry(pose,up,uc,[2,5,1],[3,6,4]),126);
  if (!out.every(Number.isFinite)) throw new Error('Invalid landmarks');
  return out;
}

export class SignSequence {
  constructor() { this.reset(); }
  reset() { this.frames=[]; this.missingSince=null; this.lastPrediction=-Infinity; }
  missing(timestamp) {
    if (this.missingSince===null) this.missingSince=timestamp;
    if (timestamp-this.missingSince>=1500) {
      this.frames=[];
      this.lastPrediction=-Infinity;
    }
  }
  add(features, timestamp) {
    // Processing latency is not loss of tracking. Only observed absence resets.
    if (this.missingSince!==null && timestamp-this.missingSince>=1500) this.reset();
    this.missingSince=null;
    this.frames.push(features);
    if (this.frames.length>30) this.frames.shift();
    if (this.frames.length<30 || timestamp-this.lastPrediction<1000) return null;
    this.lastPrediction=timestamp;
    const input=new Float32Array(30*150);
    this.frames.forEach((frame,i)=>input.set(frame,i*150));
    return input;
  }
}

export function predictionText(logits, labels) {
  if (logits.length!==112 || labels.length!==112 || !Array.from(logits).every(Number.isFinite)) {
    throw new Error('Invalid sign model output');
  }
  const best=Array.from(logits).reduce((a,_,i)=>logits[i]>logits[a]?i:a,0);
  const confidence=1/Array.from(logits).reduce((sum,v)=>sum+Math.exp(v-logits[best]),0);
  return `${labels[best]} · ${Math.round(confidence*100)}%`;
}

File: android/app/src/main/java/com/example/testsign/SignFeatures.java
package com.example.testsign;

/** Port of assets/modules/features.py; input landmarks use the unmirrored frame. */
public final class SignFeatures {
    private static final int[] HP = {0,1,2,3,0,5,6,7,0,9,10,11,0,13,14,15,0,17,18,19};
    private static final int[] HC = {1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20};
    private static final int[] UP = {1,1,2,3,1,5,6}, UC = {0,2,3,4,5,6,7};
    private SignFeatures() { }

    public static float[] extract(float[][] right, float[][] left, float[][] landmarks, int w, int h) {
        float[][] pose = new float[8][3];
        if (landmarks != null) {
            int[] indices = {0,11,12,14,16,11,13,15};
            for (int i=0;i<8;i++) pose[i] = pixel(landmarks[indices[i]], w, h);
            for (int j=0;j<3;j++) pose[1][j] = (pose[5][j]+pose[2][j])/2;
        }
        float[] neck = pose[1].clone();
        double width = 0;
        for (int j=0;j<3;j++) width += Math.pow(pose[5][j]-pose[2][j],2);
        float scale = (float)Math.sqrt(width);
        if (scale<1e-6f) scale=1;
        normalize(pose,neck,scale);
        float[][] rh = new float[21][3], lh = new float[21][3];
        if (right!=null) {
            for (int i=0;i<21;i++) rh[i]=pixel(right[i],w,h);
            normalize(rh,neck,scale);
        }
        if (left!=null) {
            for (int i=0;i<21;i++) lh[i]=pixel(left[i],w,h);
            normalize(lh,neck,scale);
        }
        float[] out=new float[150];
        geometry(rh,HP,HC,new int[]{0,4,8},new int[]{1,5,9},out,0);
        geometry(lh,HP,HC,new int[]{0,4,8},new int[]{1,5,9},out,63);
        geometry(pose,UP,UC,new int[]{2,5,1},new int[]{3,6,4},out,126);
        for (float value:out) if (!Float.isFinite(value)) throw new IllegalArgumentException("Invalid landmarks");
        return out;
    }
    private static float[] pixel(float[] p,int w,int h) { return new float[]{p[0]*w,p[1]*h,p[2]*w}; }
    private static void normalize(float[][] coords,float[] neck,float scale) {
        for (float[] p:coords) for (int j=0;j<3;j++) p[j]=(p[j]-neck[j])/scale;
    }
    private static void geometry(float[][] coords,int[] parents,int[] children,int[] a,int[] b,float[] out,int offset) {
        boolean empty=true;
        for (float[] p:coords) for(float v:p) if(v!=0) empty=false;
        if(empty) return;
        int n=parents.length;
        float[][] bones=new float[n][3];
        for(int i=0;i<n;i++) {
            for(int j=0;j<3;j++) bones[i][j]=coords[children[i]][j]-coords[parents[i]][j];
            double norm=Math.hypot(bones[i][0],bones[i][1])+1e-6;
            out[offset+i]=(float)(bones[i][0]/norm);
            out[offset+n+i]=(float)(bones[i][1]/norm);
            out[offset+2*n+i]=bones[i][2];
        }
        for(int i=0;i<a.length;i++) {
            float[] v=bones[a[i]],u=bones[b[i]];
            double dot=(double)v[0]*u[0]+(double)v[1]*u[1];
            double denom=Math.hypot(v[0],v[1])*Math.hypot(u[0],u[1])+1e-6;
            out[offset+3*n+i]=(float)Math.acos(Math.max(-1,Math.min(1,dot/denom)));
        }
    }
}
